internal/controller/workload_controller.go

```diff
@@ -90,11 +95,191 @@ func (aw *AppWrapper) PodSets() []kueue.PodSet {
 
 func (aw *AppWrapper) RunWithPodSetsInfo(podSetsInfo []podset.PodSetInfo) error {
 	aw.Spec.Suspend = false
-	return nil // TODO
+
+	// Update aw.Spec.Components to inject our labels and Kueue's PodSetInfo into every nested PodTemplateSpec
+	podSetsInfoIndex := 0
+	for componentIndex := range aw.Spec.Components {
+		component := &aw.Spec.Components[componentIndex]
+		objChanged := false
+		obj := &unstructured.Unstructured{}
+		if _, _, err := unstructured.UnstructuredJSONScheme.Decode(component.Template.Raw, nil, obj); err != nil {
+			return err
+		}
+
+		for _, podSet := range component.PodSets {
+			podSetsInfoIndex += 1
+			if podSetsInfoIndex <= len(podSetsInfo) {
+				toInject := podSetsInfo[podSetsInfoIndex-1]
+				parts := strings.Split(podSet.Path, ".")
+				p := obj.UnstructuredContent()
+				var ok bool
+				for i := 1; i < len(parts); i++ {
+					p, ok = p[parts[i]].(map[string]interface{})
+					if !ok {
+						return fmt.Errorf("path element %v not found (segment %v of %v)", parts[i], i, len(parts))
+					}
+				}
+				objChanged = true // Even if currentInfo is empty, we will still add appWrapperLabel to p.metadata.labels
+
+				if _, ok := p["metadata"]; !ok {
+					p["metadata"] = make(map[string]interface{})
```

Comment:
`MergeKeepFirst(a. b)` handles the `a == nil` case. Maybe refactor to call `HasConflict` only if `p[
"metadata"]` is defined and then unconditionally `MergeKeepFirst` without creating an empty map.

Reply:
I restructured, but it still has some conditional logic because x.(T) fails if x is nil, and we can't pass a pure interface to the `maputils` methods.